query/src/main/java/org/infinispan/query/affinity/AffinityIndexManager.java

```diff
@@ -70,37 +75,55 @@
    private final Lock readLock = flushLock.readLock();
    private ExecutorService asyncExecutor;
    private TransactionHelper transactionHelper;
-
+   private int segment;
 
    @Override
-   public void lockObtainFailed(List<LuceneWork> failingOperations) {
-      logDebug("Operation %s failed due to lock already in used", failingOperations);
+   public void operationsFailed(List<LuceneWork> failingOperations, Throwable cause) {
+      log.debugf(cause, "Operations '%s' failed in the backend", failingOperations);
+      if (cause instanceof LockObtainFailedException) {
+         this.lockObtainFailed(failingOperations);
+      }
+      if (cause instanceof SearchException) {
+         Throwable rootCause = cause.getCause();
+         if (rootCause instanceof InvalidLockException) {
+            log.warnf(cause, "Retrying failed operations");
+            this.retryOperations(failingOperations);
+         }
+      }
+   }
+
+   private void lockObtainFailed(List<LuceneWork> failingOperations) {
+      log.debugf("Operation %s failed due to lock already in used", failingOperations);
       try {
          Thread.sleep(POLL_WAIT);
       } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
       }
       Address lockHolder = getLockHolder(getIndexName());
       clearIfNeeded(lockHolder);
+      retryOperations(failingOperations);
+   }
+
+   private void retryOperations(List<LuceneWork> failingOperations) {
```

Comment:
The retries are already being done until it succeeds. Each ```performOperations``` calls involves a series of calls inside hibernate search. When an error occurs there, the callback method ```AffinityIndexManager#operationsFailed``` is called and then is retried again. Retries are done only for lock obtention related issues, so if for some reason the index is completely corrupt, it will not retry and log an error